setup.py

```diff
@@ -65,6 +66,29 @@ def run(self):
     "pulp-container>=2.3.1"
 ]
 
+
+is_on_dev_environment = (
+    "COMPOSE_PROFILE" in os.environ and "DEV_SOURCE_PATH" in os.environ
+    and os.environ.get("LOCK_REQUIREMENTS") == "0"
+)
+if is_on_dev_environment:
+    """
+    To enable the installation of local dependencies e.g: a local fork of
+    pulp_ansible checked out to specific branch/version.
+    The paths listed on DEV_SOURCE_PATH must be unpinned to avoid pip
+    VersionConflict error.
+    ref: https://github.com/ansible/galaxy_ng/wiki/Development-Setup
+         #steps-to-run-dev-environment-with-specific-upstream-branch
+    """
+    DEV_SOURCE_PATH = os.environ.get("DEV_SOURCE_PATH", "").split(":")
+    DEV_SOURCE_PATH += [path.replace("_", "-") for path in DEV_SOURCE_PATH]
+    requirements = [
+        re.sub(r"""(=|^|~|<|>|!)([\S]+)""", "", req)
+        if req.lower().startswith(tuple(DEV_SOURCE_PATH)) else req
+        for req in requirements
+    ]
+    print("Installing with unpinned DEV_SOURCE_PATH requirements", requirements)
+
```

Comment:
This is possibly somewhere we could make use of 'constraint.txt' files.  If memory servers, for our setup
we would have a requirements.txt that is all the unpinned deps. A dev build would use only requirements.txt.

For a compose profile that needs to pin versions, there would be a 'requirements/constraints.standalone.in' and .txt.
`requirements.standalone.in` would provide more specific version requires, and then  `requirements/constraints.standalone.txt` would be the resolved fully pinned requirements. 

Places where we use `pip -r` in the compose and Dockerfiles would need to use '--constraints' instead.

Reply:
@alikins the only problem is that we still need `setup.py` to hold pinned deps, so releases are based only on setup.py list, this is what is going to the RPMs build.

We can read a requirements file from setup.py, and also it is possible to make it dynamic as long as we include it as a static file on MANIFEST.in

Probably at some point we might want to move to pyproject.toml / Poetry ?